feedback.md

```diff
@@ -0,0 +1,32 @@
+Hi Kim!
+
+Many thanks for this awesome submission. I went through all (I hope) files and have a few follow-up questions - no coding needed, just conceptual stuff!
+
+Hope that's fine! :)
+
+# Worker:
+
+https://github.com/Burgestrand/kbc19/blob/master/lib/tasks/active_job.rake all good here! Clean and concise!
+
+# Jobs:
+
+https://github.com/Burgestrand/kbc19/blob/master/app/jobs/application_job.rb#L5: we may want to file this as a bug in rails/rails; if I remember correctly connection pool errors during deserialize would lead to discard!
+
+# Adapter:
+
+https://github.com/Burgestrand/kbc19/blob/master/app/jobs/google_pub_sub_adapter.rb
+
+- Maybe should go to lib? Unsure.
+- Naming: Queue makes sense to way we use it!
+- I would suggest to put execution into the worker, so the adapter only deals with enqueueing
+- https://github.com/Burgestrand/kbc19/blob/master/app/jobs/google_pub_sub_adapter.rb#L105 what would you do if execution errored out? Currently this would take the job processor down. Think of at-least-once vs at-most-once, or job duplication vs job loss.
```

Comment:
Alright, this reply is a long one so I'll TLDR first: it's [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) by design! Google handles most of our errors (1), the rest should kill our worker (2). The worker should be supervised (3) and restarted when dangerous failures occur.

1) Google catches all errors (actually only `StandardError`, more on that later) for us, and gracefully continues operations. If we want to catch these errors there's an [#on_error](https://googleapis.dev/ruby/google-cloud-pubsub/latest/Google/Cloud/PubSub/Subscriber.html#on_error-instance_method) available. In the code here we simply log it to the default rails logger, but in practice one would send it to e.g. `Sentry` in order to keep a watchful eye on increased error rates.

https://github.com/Burgestrand/kbc19/blob/c4c6938e9fedf9e4ee7a5b656f429828b79c57e5/lib/tasks/active_job.rake#L41-L43

2) Relevant code is in [Stream#background_run](https://github.com/googleapis/google-cloud-ruby/blob/296f47253a192548626682b61ff9c089f5b184a4/google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb#L285-L288). Errors that don't inherit from `StandardError` aren't caught, e.g. `OutOfMemoryError`, `Interrupt` and `SystemExit`. This is a good thing. Catching these errors (i.e. `rescue Exception`) might leave our entire Ruby process in an undefined/dangerous/scary/corrupted state at worst. These are the kinds of errors that _should_ bring our entire process down.

   Unfortunately they don't. Yet. [Thread.abort_on_exception](https://ruby-doc.org/core-2.6.1/Thread.html#method-c-abort_on_exception-3D) should be enabled by our worker before starting up, to make sure non-handled errors are fatal so that we don't end up with a workerless worker!

3) In a production setting we should run the worker using a tool designed for the job. It needs to launch our worker, restart it if it crashes (with a back off policy), and alert us if restarting does not resolve the crashing based on some heuristic. An example of this would be [systemd](https://www.freedesktop.org/software/systemd/man/systemd.service), which is available on most distributions and is basically built for this.

Reply:
And finally! When it comes to at-most-once or at-least-once, [google guarantees at-least-once](https://cloud.google.com/pubsub/docs/faq#duplicates) and I'd argue that's the policy that makes the most sense in most cases.

In short, I'd argue for always defaulting to idempotent workers if at all possible.

If absolutely needed, we can always implement at-most-once with at-least-once workers, e.g. using an external database to keep track of jobs that have been started on, and making sure our workers wait to verify their jobs before starting them.